pkg/storegateway/series_refs.go

```diff
@@ -64,36 +71,41 @@ func (m seriesChunkRef) Compare(other seriesChunkRef) int {
 	return 0
 }
 
-// seriesChunkRefsIterator implements an interator over seriesChunkRefsSet.
-type seriesChunkRefsIterator struct {
+// seriesChunkRefsIteratorImpl implements an iterator returning a sequence of seriesChunkRefs.
+type seriesChunkRefsIteratorImpl struct {
 	currentOffset int
 	set           seriesChunkRefsSet
 }
 
-func newSeriesChunkRefsIterator(set seriesChunkRefsSet) *seriesChunkRefsIterator {
-	return &seriesChunkRefsIterator{
-		set:           set,
-		currentOffset: -1,
-	}
+func newSeriesChunkRefsIterator(set seriesChunkRefsSet) *seriesChunkRefsIteratorImpl {
+	c := &seriesChunkRefsIteratorImpl{}
+	c.reset(set)
+
+	return c
 }
 
-// reset replaces the current set with the provided one. There is no need to call Next() after reset().
-func (c *seriesChunkRefsIterator) reset(set seriesChunkRefsSet) {
+// reset replaces the current set with the provided one. After calling reset() you
+// must call Next() to advance the iterator to the first element.
+func (c *seriesChunkRefsIteratorImpl) reset(set seriesChunkRefsSet) {
 	c.set = set
-	c.currentOffset = 0
+	c.currentOffset = -1
 }
 
-func (c *seriesChunkRefsIterator) Next() bool {
+func (c *seriesChunkRefsIteratorImpl) Next() bool {
 	c.currentOffset++
 	return !c.Done()
 }
 
-func (c *seriesChunkRefsIterator) Done() bool {
-	return c.currentOffset < 0 || c.currentOffset >= c.set.len()
+// Done returns true if the iterator trespassed the end and the item returned by At()
+// is the zero value. If the iterator is on the last item, the value returned by At()
+// is the actual item and Done() returns false.
+func (c *seriesChunkRefsIteratorImpl) Done() bool {
+	setLength := c.set.len()
+	return setLength == 0 || c.currentOffset >= setLength
 }
 
-func (c *seriesChunkRefsIterator) At() seriesChunkRefs {
-	if c.Done() {
+func (c *seriesChunkRefsIteratorImpl) At() seriesChunkRefs {
+	if c.currentOffset < 0 || c.currentOffset >= c.set.len() {
 		return seriesChunkRefs{}
 	}
 	return c.set.series[c.currentOffset]
```

Comment:
```suggestion
	return c.set.series[c.currentOffset]
}
func (*seriesChunkRefsIteratorImpl) Err() error { 
	return nil
```

Reply:
ok with naming. but then let's make it actually implement `seriesChunkRefsIterator`

Reply:
Eh eh good point! Done in https://github.com/grafana/mimir/pull/3585/commits/1f069868dbf9fcb2562d6cb60d3dd4474a443e70.